diff --git a/crates/storage/libmdbx-rs/mdbx-sys/build.rs b/crates/storage/libmdbx-rs/mdbx-sys/build.rs
index c265d02e23..b45672a8e4 100644
--- a/crates/storage/libmdbx-rs/mdbx-sys/build.rs
+++ b/crates/storage/libmdbx-rs/mdbx-sys/build.rs
@@ -39,6 +39,7 @@ fn main() {
     }
 
     cc.file(mdbx.join("mdbx.c")).compile("libmdbx.a");
+    println!("cargo:rustc-link-lib=uring");
 }
 
 fn generate_bindings(mdbx: &Path, out_file: &Path) {
