// @openzeppelin/upgrades-core@1.34.1/src/index.ts
export * from './validate';
export { fetchOrDeploy, fetchOrDeployAdmin, fetchOrDeployGetDeployment } from './impl-store';
export * from './version';
export * from './storage';
export {
  EIP1967BeaconNotFound,
  EIP1967ImplementationNotFound,
  getAdminAddress,
  getBeaconAddress,
  getImplementationAddress,
  toEip1967Hash,
  toFallbackEip1967Hash,
  isEmptySlot,
} from './eip-1967';
export * from './provider';
export * from './src-decoder';
export * from './solc-api';
export * from './deployment';
export * from './link-refs';
export * from './error';

export {
  ManifestData,
  ImplDeployment,
  ProxyDeployment,
  Manifest,
  migrateManifest,
  DeploymentNotFound,
} from './manifest';

export { getStorageLayoutForAddress } from './manifest-storage-layout';

export * from './scripts/migrate-oz-cli-project';

export { logWarning } from './utils/log';
export { setProxyKind, processProxyKind, detectProxyKind } from './proxy-kind';

export { UpgradeableContract } from './standalone';

export { isTransparentOrUUPSProxy, isBeaconProxy, isTransparentProxy } from './eip-1967-type';
export { getImplementationAddressFromBeacon, getImplementationAddressFromProxy } from './impl-address';
export { isBeacon } from './beacon';
export { addProxyToManifest } from './add-proxy-to-manifest';

export {
  BeaconProxyUnsupportedError,
  LoadProxyUnsupportedError,
  PrepareUpgradeUnsupportedError,
  DeployBeaconProxyUnsupportedError,
  DeployBeaconProxyImplUnknownError,
  DeployBeaconProxyKindError,
  ForceImportUnsupportedError,
  NoContractImportError,
  ValidateUpdateRequiresKindError,
  PrepareUpgradeRequiresKindError,
  InitialOwnerUnsupportedKindError,
  assertNotProxy,
} from './usage-error';

export { ValidateUpgradeSafetyOptions, validateUpgradeSafety, ProjectReport, ReferenceContractNotFound } from './cli';

export { getUpgradeInterfaceVersion } from './upgrade-interface-version';
export { makeNamespacedInput } from './utils/make-namespaced';
export { isNamespaceSupported } from './storage/namespace';
