diff --git a/internal/libhive/run.go b/internal/libhive/run.go
index b012296..c745745 100644
--- a/internal/libhive/run.go
+++ b/internal/libhive/run.go
@@ -43,13 +43,10 @@ func NewRunner(inv Inventory, b Builder, cb ContainerBackend) *Runner {
 
 // Build builds client and simulator images.
 func (r *Runner) Build(ctx context.Context, clientList []ClientDesignator, simList []string) error {
-	if err := r.container.Build(ctx, r.builder); err != nil {
-		return err
-	}
 	if err := r.buildClients(ctx, clientList); err != nil {
 		return err
 	}
-	return r.buildSimulators(ctx, simList)
+	return nil
 }
 
 // buildClients builds client images.
